### PR TITLE
Migrate to the latest libp2p

### LINKED ISSUE #629: Update js-libp2p to 0.27
See libp2p/js-libp2p#487

### PR DESCRIPTION
**Pull request by twoeths**
## Goal
+ Close #629

### CODE REVIEW

File: packages/lodestar-cli/tsconfig.json [hunk 1]
```diff
@@ -4,7 +4,7 @@
   "compilerOptions": {
     "typeRoots": [
       "../../node_modules/@types",
-      "../../node_modules/@polkadot/ts"
```
Comment (twoeths) [inline packages/lodestar-cli/tsconfig.json:7]: probably we want a separate module, or just fork polkadot-ts?

Comment (mpetrunic) [reply packages/lodestar-cli/tsconfig.json:7]: I think we should fork polkadot-ts and rename

File: packages/lodestar-cli/tsconfig.json [hunk 2]
```diff
@@ -4,7 +4,7 @@
...
+      "../../packages/lodestar/src/@types"
```
Comment (mpetrunic) [inline packages/lodestar-cli/tsconfig.json]: wouldn't ./src/@types be safer?

File: packages/lodestar/src/@types/interface-connection/index.d.ts [hunk 3]
```diff
@@ -0,0 +1,10 @@
+// Type definitions for interface-connection 0.3.2
+// Project: https://github.com/libp2p/interface-connection
+// Definitions by: Jaco Greeff <https://github.com/jacogr>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+
+/// <reference types="peer-info"/>
+
+declare interface LibP2pConnection {
+  getPeerInfo (cb: (error: Error | null, peerInfo?: PeerInfo) => any): void;
```
Comment (mpetrunic) [inline packages/lodestar/src/@types/interface-connection/index.d.ts]: any => void

File: packages/lodestar/src/@types/libp2p-bootstrap/index.d.ts [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+    interval?: number;
+  };
+
+  type Events = "peer";
+}
+
+declare class LibP2pBootstrap {
+  constructor (options: LibP2pBootstrap.Options);
+
+  on (event: LibP2pBootstrap.Events, cb: (peerInfo: PeerInfo) => any): this;
```
Comment (mpetrunic) [inline packages/lodestar/src/@types/libp2p-bootstrap/index.d.ts]: Probably any => void

File: packages/lodestar/src/@types/peer-id/index.d.ts [hunk 5]
```diff
@@ -0,0 +1,43 @@
+// Type definitions for peer-id 0.10.7
+// Project: https://github.com/libp2p/js-peer-id
+// Definitions by: Jaco Greeff <https://github.com/jacogr>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+
+/// <reference types="node"/>
+/// <reference types="libp2p-crypto"/>
+
+declare namespace PeerId {
```
Comment (mpetrunic) [inline packages/lodestar/src/@types/peer-id/index.d.ts]: This package has types already

Comment (mpetrunic) [reply packages/lodestar/src/@types/peer-id/index.d.ts]: You should check if some other lib already has types

File: packages/lodestar/src/@types/pull-mplex/index.d.ts [hunk 6]
```diff
@@ -0,0 +1,13 @@
+// Type definitions for pull-mplex 0.1.1
+// Project: https://github.com/libp2p/pull-mplex
+// Definitions by: Jaco Greeff <https://github.com/jacogr>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+
+declare interface PullMplex {
```
Comment (mpetrunic) [inline packages/lodestar/src/@types/pull-mplex/index.d.ts]: Isn't this pull stream related?

File: packages/lodestar/src/network/gossip/gossip.ts [hunk 7]
```diff
@@ -46,19 +45,20 @@ export class Gossip extends (EventEmitter as { new(): GossipEventEmitter }) impl
     this.logger = logger.child({module: "gossip", level: LogLevel[logger.level]});
     this.logger.silent = logger.silent;
     this.validator = validator;
-    this.pubsub = new Gossipsub(libp2p, {gossipIncoming: false});
+    // @ts-ignore
+    this.pubsub = new Gossipsub(libp2p.peerInfo, libp2p.registrar, {gossipIncoming: false});
```
Comment (mpetrunic) [inline packages/lodestar/src/network/gossip/gossip.ts:48]: When you are at it, might be good idea to add gossipsub types which we can move to it's repo later on?

File: packages/lodestar/src/network/network.ts [hunk 8]
```diff
@@ -60,43 +59,51 @@ export class Libp2pNetwork extends (EventEmitter as { new(): NetworkEventEmitter
...
-      return false;
-    }
-    return Boolean(peer.isConnected());
+    // @ts-ignore
+    return !!this.libp2p.registrar.getConnection(peerInfo);
   }
+
   public async connect(peerInfo: PeerInfo): Promise<void> {
-    await promisify<void, PeerInfo>(this.libp2p.dial.bind(this.libp2p))(peerInfo);
+    // @ts-ignore
```
Comment (mpetrunic) [inline packages/lodestar/src/network/network.ts]: What's up with all those ts-ignores?

File: packages/lodestar/src/network/reqResp.ts [hunk 9]
```diff
@@ -83,73 +94,53 @@ export class ReqResp extends (EventEmitter as IReqRespEventEmitterClass) impleme
...
+        this.removeListener(responseEvent, responseListener);
+        resolve(this.encodeResponseError(new Error(ERR_RESP_TIMEOUT)));
+      }, RESP_TIMEOUT);
+      responseListener = (err: Error|null, output: ResponseBody): void => {
+        // @ts-ignore
+        this.removeListener(responseEvent, responseListener);
+        clearTimeout(responseTimer);
+        if (err) resolve(this.encodeResponseError(err));
+        this.logger.verbose(`${requestId} - send ${method} response`);
+        resolve(this.encodeResponse(method, output));
```
Comment (mpetrunic) [inline packages/lodestar/src/network/reqResp.ts:142]: Would be nice while you are at it to rewrite this method cleaner withot ts-ignores and eslint disables

**Review by mpetrunic — CHANGES_REQUESTED** [review]
Few things:
- fix types (add missing, remove duplicates)
- fix all those ts-ignores, typescript is useless if we are gonna ignore everything, we could have bugs there without knowing

File: packages/lodestar/src/@types/peer-id/index.d.ts (hunk 5 above)
Comment (mpetrunic) [reply packages/lodestar/src/@types/peer-id/index.d.ts]: Like libp2p-crypto

File: packages/lodestar-cli/tsconfig.json (hunk 2 above)
Comment (twoeths) [reply packages/lodestar-cli/tsconfig.json]: this is in lodestar-cli, I referred to lodestar package.
Should be clean if we move all types to libp2p-ts.

**Comment by twoeths** [comment]
Updating this PR with new libp2p-ts

File: packages/lodestar/src/@types/pull-mplex/index.d.ts (hunk 6 above)
Comment (twoeths) [reply packages/lodestar/src/@types/pull-mplex/index.d.ts]: yeah I guess so. Anyway it's removed in the new version of libp2p, done in libp2p-ts

File: packages/lodestar/src/@types/interface-connection/index.d.ts (hunk 3 above)
Comment (twoeths) [reply packages/lodestar/src/@types/interface-connection/index.d.ts]: this is obsolete, will remove it

File: packages/lodestar/src/network/network.ts (hunk 8 above)
Comment (twoeths) [reply packages/lodestar/src/network/network.ts]: will find a way to clean it

File: packages/lodestar/src/network/gossip/gossip.ts (hunk 7 above)
Comment (twoeths) [reply packages/lodestar/src/network/gossip/gossip.ts:48]: I have a separate PR to deal with new gossip, will work on that in that PR

**Comment by twoeths** [comment]
Ready for review

File: packages/lodestar/src/network/reqResp.ts [hunk 10]
```diff
@@ -245,52 +241,37 @@ export class ReqResp extends (EventEmitter as IReqRespEventEmitterClass) impleme
     }
   }
   private async sendRequest<T extends ResponseBody>(
-    peerInfo: PeerInfo,
+    peerId: PeerId,
```
Comment (wemeetagain) [inline packages/lodestar/src/network/reqResp.ts]: Can we keep this as `PeerInfo`?
It looks like `dialProtocol`, below should work with a PeerInfo too. https://github.com/libp2p/js-libp2p/blob/master/src/index.js#L282

**Review by wemeetagain — CHANGES_REQUESTED** [review]

File: packages/lodestar-cli/package.json [hunk 11]
```diff
@@ -54,7 +54,7 @@
     "deepmerge": "^3.2.0",
     "ethers": "^4.0.40",
     "js-yaml": "^3.13.1",
-    "peer-id": "^0.12.2",
+    "peer-id": "^0.13.4",
```
Comment (mpetrunic) [inline packages/lodestar-cli/package.json]: I think types are released in version 0.13.7

File: packages/lodestar/src/network/gossip/interface.ts [hunk 12]
```diff
@@ -38,9 +38,9 @@ export interface IGossipModules {
 }
 
 export interface IGossipSub extends EventEmitter {
-  publish(topic: string, data: Buffer, cb: (err: unknown) => void): void;
-  start(cb: (err: unknown) => void): void;
-  stop(cb: (err: unknown) => void): void;
+  publish(topic: string, data: Buffer): void;
+  start(): void;
+  stop(): void;
```
Comment (mpetrunic) [inline packages/lodestar/src/network/gossip/interface.ts]: I think all three methods returns Promise now

**Review by mpetrunic — COMMENTED** [review]
much better

File: packages/lodestar-cli/package.json (hunk 11 above)
Comment (twoeths) [reply packages/lodestar-cli/package.json]: yeah really

File: packages/lodestar/src/network/reqResp.ts (hunk 10 above)
Comment (twoeths) [reply packages/lodestar/src/network/reqResp.ts]: yeah I changed it back to PeerInfo across interfaces.

**Review by mpetrunic — APPROVED** [review]

**Review by wemeetagain — APPROVED** [review]

**Pull request merged**